pkg/services/auth/authimpl/external_session_store.go

```diff
@@ -150,6 +150,45 @@ func (s *store) Create(ctx context.Context, extSession *auth.ExternalSession) er
 	return nil
 }
 
+func (s *store) Update(ctx context.Context, ID int64, cmd *auth.UpdateExternalSessionCommand) error {
+	ctx, span := s.tracer.Start(ctx, "externalsession.Update")
+	defer span.End()
+
+	var err error
+	externalSession := &auth.ExternalSession{}
+
+	externalSession.AccessToken, err = s.encryptAndEncode(cmd.Token.AccessToken)
+	if err != nil {
+		return err
+	}
+
+	externalSession.RefreshToken, err = s.encryptAndEncode(cmd.Token.RefreshToken)
+	if err != nil {
+		return err
+	}
+
+	var secretIdToken string
+	if idToken, ok := cmd.Token.Extra("id_token").(string); ok && idToken != "" {
+		secretIdToken, err = s.encryptAndEncode(idToken)
+		if err != nil {
+			return err
+		}
+		externalSession.IDToken = secretIdToken
+	}
+
+	externalSession.ExpiresAt = cmd.Token.Expiry
+
+	err = s.sqlStore.WithDbSession(ctx, func(sess *db.Session) error {
+		_, err := sess.ID(ID).Cols("access_token", "refresh_token", "id_token", "expires_at").Update(externalSession)
```

Comment:
Some Qs on this one:
- Do we have an `updated_at` field we want to update here with the current timestamp?
- I remember we also have hashes for secrets, but I don't remember whether we keep them in the DB. Just checking if we have to update the hashes too.

Reply:
> Do we have an updated_at field we want to update here with the current timestamp?

Nope, we don't have that, only `created_at` and `expires_at`. We can add that, however I'm not sure it would provide much value, because the expiry is stored and the `user_external_session` entry gets removed when the Grafana session/external session ends.

> I remember we also have hashes for secrets, but I don't remember whether we keep them in the DB. Just checking if we have to update the hashes too.

Yes, we store the hash for the `NameID` (only for SAML) and the `SessionID` (applies to OAuth as well), but the `SessionID` should not change on a token refresh.